### PR TITLE
ROX-22289: drop kube-rbac-proxy sidecar

### PR DESCRIPTION
**Pull request by porridge**
## Description

Its functionality is now built into the controller-runtime library, so the manager can serve the metrics endpoint directly. This lets us get rid of a dependency that proved to be problematic (see ticket's parent for more background) and make a tiny dent in operator pod's RAM/CPU needs.

AFAICT the `go.mod` changes are a result of importing "sigs.k8s.io/controller-runtime/pkg/metrics/filters" which has additional dependencies.

I hope making `--replace-rbac-proxy` in the `patch-csv.py` script into a no-op is a reasonable migration strategy, @msugakov ? Additional downstream cleanup will be tracked in [ROX-22355](https://issues.redhat.com/browse/ROX-22355).

## Checklist
- [x] Investigated and inspected CI test results
- [x] ~Unit test and regression tests added~ - tested by tests added in #9773 
- [x] ~Evaluated and added CHANGELOG entry if required~
- [x] ~Determined and documented upgrade steps~
- [x] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

This change should be transparent to users.

### Here I tell how I validated my change

CI should be enough with the recently added metrics test mentioned above.

As requested by @msugakov I also did a manual test following [these instructions](https://docs.openshift.com/container-platform/4.12/operators/operator_sdk/osdk-monitoring-prometheus.html) (`make manifests bundle` after uncommenting the prometheus line, to obtain the `ServiceMonitor` definition in step one).

This works as expected:

Screenshot from 2024-02-08 15-57-46

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Review by kovayur — APPROVED** [review]

**Review by msugakov — COMMENTED** [review]
Thank you, this is great!

Having tests https://github.com/stackrox/stackrox/pull/9773/ created before the change and used to test the change is even more awesome.

That said, I wasn't able to fully draw a parallel between tests that you implemented and what I had to do to see metrics flowing in OCP built-in Prometheus. I described that [here](https://redhat-internal.slack.com/archives/C01Q93MMCNT/p1674838988734469).
Do you mind trying that as a manual test to confirm the metrics appear in OCP console?

File: operator/config/default/kustomization.yaml [hunk 1]
```diff
@@ -27,7 +27,8 @@ patchesStrategicMerge:
 # Protect the /metrics endpoint by putting it behind auth.
 # If you want your controller-manager to expose the /metrics
 # endpoint w/o any authn/z, please comment the following line.
-- manager_auth_proxy_patch.yaml
+# The authentication is now done in-process, by controller-runtime.
+#- manager_auth_proxy_patch.yaml
```
[the remark this answers is not in the document]
Comment (porridge) [reply operator/config/default/kustomization.yaml:31]: My intention was to reduce the number of conflicting lines and leave an explicit hint to @ludydoo who is working in parallel on bumping the scaffolding ([ROX-17208](https://issues.redhat.com/browse/ROX-17208)).

File: operator/config/manager/manager.yaml [hunk 2]
```diff
@@ -32,6 +32,8 @@ spec:
         runAsNonRoot: true
       containers:
       - args:
+        - "--health-probe-bind-address=:8081"
+        - "--metrics-bind-address=0.0.0.0:8443"
```
[the remark this answers is not in the document]
Comment (porridge) [reply operator/config/manager/manager.yaml:36]: Yeah, the problem is that I'm not sure they mean the same. I would actually expect that empty IP means localhost-only, so I wanted to make sure the metrics endpoint works as expected.

File: operator/config/manager/manager.yaml [hunk 3]
```diff
@@ -52,6 +54,10 @@ spec:
         name: manager
         securityContext:
           allowPrivilegeEscalation: false
+        ports:
+        - containerPort: 8443
+          protocol: TCP
+          name: https
```
[the remark this answers is not in the document]
Comment (ludydoo) [reply operator/config/manager/manager.yaml:60]: nit: A nice convention I used from the past from istio (that allows automatic protocol selection) is to prefix `<protocol>-` to the port name, e.g. `https-metrics` or `tcp-postgres`

[the remark this answers is not in the document]
Comment (porridge) [reply operator/config/manager/manager.yaml:60]: There was and is a service for the metrics port. That's what the aforementioned [tests are using](https://github.com/stackrox/stackrox/blob/5e5118ef2a53ce288d19f72026ab49145fd48973/operator/tests/controller/metrics/200-access-privileged.yaml#L20) for example. I'm not sure what the side-effects of changing the port name would be, so I'd rather keep it as is.

File: operator/main.go [hunk 4]
```diff
@@ -128,14 +129,34 @@ func run() error {
 	}
 	defer restore()
 
+	var tlsOpts []func(c *tls.Config)
+	if !enableHTTP2 {
```
[the remark this answers is not in the document]
Comment (porridge) [reply operator/main.go:133]: Yes, the third argument [here](https://github.com/stackrox/stackrox/blob/5e5118ef2a53ce288d19f72026ab49145fd48973/operator/main.go#L115) is the default, which is a bool var without explicit initialization, so `false`.

[the remark this answers is not in the document]
Comment (vladbologa) [reply operator/main.go:133]: It's a bit confusing indeed. This code is coming from an operator remediation guideline doc that was distributed by the operator framework team.

**Pull request merged**